src/lambda_toolshed/papillon.cljc

```diff
@@ -23,48 +23,44 @@
   #?(:clj (instance? Throwable x)
      :cljs (instance? js/Error x)))
 
-(defn- async-catch
-  "Takes a value from the channel `c` and checks if it is an error type value.
-  If the result is an error type value then add that to the context `ctx` under
-  the `:lambda-toolshed.papillon/error` key and return it, otherwise return the
-  value unchanged."
-  [ctx c]
-  (go
-    (let [x (<! c)]
-      (cond
-        (nil? x) (assoc ctx ::error (ex-info "Context channel was closed." {::ctx ctx}))
-        (error? x) (assoc ctx ::error x)
-        :else x))))
+(defn clear-queue
+  "Empty the interceptor queue of the given context `ctx`, thus ensuring no
+  further processing of the `enter` chain is attempted."
+  [ctx]
+  (update ctx ::queue empty))
+
+(defn- transition
+  "Transition the context `ctx` to the candidate context value `candidate`.  This function
+  works synchronously with value candidates -any async processing should be performed prior
+  to invoking this function."
+  [ctx candidate]
+  (cond
+    (reduced? candidate) (clear-queue (transition ctx (unreduced candidate)))
+    (error? candidate) (assoc (clear-queue ctx) ::error candidate)
+    (not= (-> candidate meta :type) ::ctx) (assoc (clear-queue ctx)
+                                                  ::error (ex-info "Context was lost!"
+                                                                   {::ctx ctx ::candidate-ctx candidate}))
+    :else candidate))
```

Comment:
how about structuring the different branches the same way?

```clojure
(defn- transition
  "Transition the context `ctx` to the candidate context value `candidate`.  This function
  works synchronously with value candidates -any async processing should be performed prior
  to invoking this function."
  [ctx candidate]
  (cond
    (-> candidate reduced?)
    (-> ctx
        (transition (unreduced candidate))
        (clear-queue))

    (-> candidate error?)
    (-> ctx
        (clear-queue)
        (assoc ::error candidate))

    (-> candidate meta :type (not= ::ctx))
    (-> ctx
        (clear-queue)
        (assoc ::error (ex-info "Context was lost!"
                                {::ctx           ctx
                                 ::candidate-ctx candidate})))
    :else candidate))
```

it clearly communicates that
1. all branch conditions are functions of the `candidate` ctx
2. all branch values are transformations of the current `ctx`

it also make the `::candidate-ctx candidate` asymmetry stand out, so i'd recommend renaming the `candidate` arg to `candidate-ctx`.

Reply:
it might worth putting the `(clear-queue)` at the end of the transformations too, so it's even more obvious, that all branches share that common transform and there is no significance in doing the clearing 1st on 2 branches, but clearing last on the 1st branch of `cond`.

Reply:
@onetom, Out of curiosity, what put papillon on your radar?

Reply:
i think, i saw first papillon being mentioned here:
https://clojurians.slack.com/archives/CDG3YQV6Z/p1651552940936609

but can't remember who pointed me towards the #interceptors clojurians slack channel.

i'm thinking about making a amalgam of these 2 libraries:
1. https://github.com/gnarroway/hato
2. https://github.com/oliyh/martian

`hato` uses middlewares and `martian` uses the ancient `tripod` (https://github.com/frankiesardo/tripod) interceptor library.

in a server context we also have `reitit`, which uses the `sieppari` (https://github.com/metosin/sieppari) interceptor library.

it would be a significant drop in mental overhead to just use a common library across all these application areas.
it would allow us to use the same tracing, diffing, chain editing approaches across the board too.

the main attraction of interceptors to me is the "ease" of testability.
if we can concentrate the side-effects into a single interceptor, then it's enough to just swap that out to a fake one (in a test-double sense), in automated tests, similarly how martian recommends it:
https://github.com/oliyh/martian#testing-with-martian-test

Reply:
That is a good description of the raison d'être of papillon.